modules/pricer/src/main/java/com/opengamma/strata/pricer/fx/ImpliedTrinomialTreeFxSingleBarrierOptionProductPricer.java

```diff
@@ -0,0 +1,470 @@
+   * <p>
+   * The present value of the product is the value on the valuation date. 
+   * It is expressed in the counter currency.
+   * <p>
+   * This assumes the tree is already calibrated and the tree data is stored as {@code RecombiningTrinomialTreeData}.
+   * The tree data should be consistent with the pricer and other inputs, see {@link #validateData}.
+   * 
+   * @param option  the option product
+   * @param ratesProvider  the rates provider
+   * @param volatilityProvider  the Black volatility provider
+   * @param treeData  the trinomial tree data
+   * @return the present value of the product
+   */
+  public CurrencyAmount presentValue(
+      ResolvedFxSingleBarrierOption option,
+      RatesProvider ratesProvider,
+      BlackVolatilityFxProvider volatilityProvider,
+      RecombiningTrinomialTreeData treeData) {
+
+    double price = price(option, ratesProvider, volatilityProvider, treeData);
+    ResolvedFxVanillaOption underlyingOption = option.getUnderlyingOption();
+    return CurrencyAmount.of(underlyingOption.getCounterCurrency(), signedNotional(underlyingOption) * price);
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Calculates the present value sensitivity of the FX barrier option product.
+   * <p>
+   * The present value sensitivity of the product is the sensitivity of {@link #presentValue} to
+   * the underlying curve parameters.
+   * <p>
+   * The sensitivity is computed by bump and re-price.
+   * 
+   * @param option  the option product
+   * @param ratesProvider  the rates provider
+   * @param volatilityProvider  the Black volatility provider
+   * @return the present value of the product
+   */
+  public CurveCurrencyParameterSensitivities presentValueCurveParameterSensitivity(
+      ResolvedFxSingleBarrierOption option,
+      RatesProvider ratesProvider,
+      BlackVolatilityFxProvider volatilityProvider) {
+
+    RecombiningTrinomialTreeData baseData = calibrateTrinomialTree(option, ratesProvider, volatilityProvider);
+    return presentValueCurveParameterSensitivity(option, ratesProvider, volatilityProvider, baseData);
+  }
+
+  /**
+   * Calculates the present value sensitivity of the FX barrier option product.
+   * <p>
+   * The present value sensitivity of the product is the sensitivity of {@link #presentValue} to
+   * the underlying curve parameters.
+   * <p>
+   * The sensitivity is computed by bump and re-price.
+   * 
+   * @param option  the option product
+   * @param ratesProvider  the rates provider
+   * @param volatilityProvider  the Black volatility provider
+   * @param baseTreeData  the trinomial tree data
+   * @return the present value of the product
+   */
+  public CurveCurrencyParameterSensitivities presentValueCurveParameterSensitivity(
+      ResolvedFxSingleBarrierOption option,
+      RatesProvider ratesProvider,
+      BlackVolatilityFxProvider volatilityProvider,
+      RecombiningTrinomialTreeData baseTreeData) {
+
```

Comment:
This method should call a generic FD calculator. The current one does not offer the possibility to restrict the number of curves bumps. Issue #988 has been created to improve on that restriction. No change are required for this comment at this stage, but it should be revisited when #988 is closed.